core/trino-main/src/main/java/io/trino/metadata/Metadata.java

```diff
@@ -347,6 +348,16 @@ Optional<ConnectorOutputMetadata> finishRefreshMaterializedView(
             Collection<ComputedStatistics> computedStatistics,
             List<TableHandle> sourceTableHandles);
 
+    /**
+     * Push update into connector
+     */
+    Optional<TableHandle> applyUpdate(Session session, TableHandle tableHandle);
```

Comment:
Don't we need assignments here? I think there could be a case where predicate is easy and assignments are complex to be pushed down. Is this handled somewhere else?

Reply:
As far as I understand this code executed only in PushMergeWriterUpdateIntoConnector rule, where we catch only concrete node plan for simple cases. So in case of complex assignment like Update with Select, we never fall here.
And flow will fail on beginMerge

Reply:
Updates cannot be pushed into a regular table scan. They have fundamentally different semantics -- scans are side-effect-free, while updates have side effects. This is important for retries and transactionality.

Reply:
We are already doing this for DELETE. 

UPDATE, like DELETE, has side effects but it is idempotent operation (as long query is deterministic). It sounds like a good idea that we should not push down DELETE and UPDATE if they are not deterministic.

Reply:
What's the decision here?

Reply:
I think the update can be retry as long as the condition is not involved in the update assignments.

Reply:
> Updates cannot be pushed into a regular table scan.

@martint notice that we are pushing update into `TableUpdateNode`, then have dedicated operator that executes the update operation. So there is no table scan here.